File: frontend/services/notification.service.ts
import api from "@/lib/api";


export interface Notification {

    id: number;

    title: string;

    description: string;

    type: string;

    is_read: boolean;

    created_at: string;

}


export interface NotificationResponse {

    items: Notification[];

    unread_count: number;

}


export async function getNotifications(): Promise<NotificationResponse> {

    const { data } = await api.get(
        "/notifications"
    );

    return data;

}


export async function markNotificationAsRead(
    notificationId: number
) {

    const { data } = await api.post(
        `/notifications/${notificationId}/read`
    );

    return data;

}


export async function markAllNotificationsAsRead() {

    const { data } = await api.post(
        "/notifications/read-all"
    );

    return data;

}

File: backend/app/services/notification_service.py
from app.entities.notification_entity import NotificationEntity
from app.repositories.notification_repository import NotificationRepository


class NotificationService:

    def __init__(self):

        self.repository = NotificationRepository()

    def create(
        self,
        user_id: int,
        title: str,
        description: str,
        type: str
    ):

        entity = NotificationEntity(

            user_id=user_id,

            title=title,

            description=description,

            type=type

        )

        return self.repository.save(entity)

    def get_notifications(
        self,
        user_id: int
    ):

        notifications = (
            self.repository.get_all_for_user(
                user_id
            )
        )

        unread_count = len([
            notification
            for notification in notifications
            if not notification.is_read
        ])

        return {
            "items": notifications,
            "unread_count": unread_count
        }

    def mark_as_read(
        self,
        notification_id: int,
        user_id: int
    ):

        return self.repository.mark_as_read(
            notification_id,
            user_id
        )

    def mark_all_as_read(
        self,
        user_id: int
    ):

        return self.repository.mark_all_as_read(
            user_id
        )